### DISCUSSION TITLE
3.2 单调栈

### DISCUSSION THREAD

**Discussion by giscus** [itcharge/AlgoNote#185] [Comments]
# 3.2 单调栈

单调栈知识 1. 单调栈简介 单调栈（Monotone Stack）：一种特殊的栈。在栈的「先进后出」规则基础上，要求「从 栈顶 到 栈底 的元素是单调递增（或者单调递减）」。其中满足从栈顶到栈底的元素是单调递增的栈，叫做「单调递增栈」。满足从栈顶到栈底的元素是单调递减的栈，叫做「单调递减栈」。 注意：这里定义的顺序是从「栈顶」到「栈底」。有的文章里是...

[https://algo.itcharge.cn/03_stack_queue_hash_table/03_02_monotone_stack/](https://algo.itcharge.cn/03_stack_queue_hash_table/03_02_monotone_stack/)

**Comment by zh820952173** [comment]
你的总结我服了，牛逼啊 单调栈的相关题目都秒了，但是对于找右侧最值时，我还是坚持使用从右到左构造单调栈，这样的话使用方式就全统一了，当前元素插入栈前进行处理栈顶（目标）元素

**Comment by xiaos2021** [comment]
向大家推荐一下，我在Leetcode写的单调栈模板题解，超级详细，实现了上面提到的所有情况的模板代码。

[单调栈的四类模板（模板框架代码，接雨水的处理，多种解法+写法，详细分析）](https://leetcode.cn/problems/trapping-rain-water/solutions/3801439/dan-diao-zhan-de-san-lei-mo-ban-mo-ban-k-fh1j/)
